include/tscore/ink_assert.h

```diff
@@ -53,13 +53,50 @@ inkcoreapi void _ink_assert(const char *a, const char *f, int l) TS_NORETURN;
 
 #ifdef __cplusplus
 }
-#endif /* __cplusplus */
 
-/* workaround a bug in the  stupid Sun preprocessor
+/*
+Use dbg_dyn_cast() to do pointer/reference dynamic upcasts with checked dynamic_cast in debug builds, and with
+static_cast in release (optimized) builds.  Use examples:
+
+class A { public: virtual ~A(); };
+class B : public A {};
+B * foo(A *a) { return dbg_dyn_cast<B>(a); }
+B & foo2(A &a) { return dbg_dyn_cast<B>(a); }
+B const & foo3(A const &a) { return dbg_dyn_cast<B const>(a); }
 
-#undef assert
-#define assert __DONT_USE_BARE_assert_USE_ink_assert__
-#define _ASSERT_H
-#undef __ASSERT_H__
-#define __ASSERT_H__
 */
+
+template <class Derived, class Base>
+Derived *
+dbg_dyn_cast(Base *b)
```

Comment:
The behavior seems more like static cast because it doesn't return nullptr. We can use this as a replacement for static_cast but not dynamic_cast.

Reply:
Hmm I was thinking this would only be used in cases where b was expected to never be null.  Probably I should have noted that in the comments.

Reply:
> I was thinking this would only be used in cases where b was expected to never be null.

I think I understand where to use this. I was trying to say the name is confusing. People use static_cast in those cases, like the code for non-debug build does, and the behavior is more like static_cast since it never returns null, but it has dyn(amic) in its name. Why do we want to call it a kind of dynamic_cast?

I'd suggest `dbg_static_cast` or something doesn't mention dynamic nor static.

Reply:
ok